### PR TITLE
Sound record module

### LINKED ISSUE #45: Add audio recording module capability to apra pipes
**Is your feature request related to a problem? Please describe.**
This feature is to allow audio frames to be recording via a microphone for processing.

**Describe the solution you'd like**
Use a third party library [sfml](https://www.sfml-dev.org/) to record audio frames from a microphone

**Describe alternatives you've considered**
No alternatives

**Additional context**
Sample rate, bit depth,number of channels etc should be configurable.

### PR DESCRIPTION
**Pull request by mansirthd**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #45 

**Description**
Added 16 bit, multi channel (mono / stereo) sound recording module. 

**Alternative(s) considered**

**Type**

Feature

**Screenshots (if applicable)**

**Checklist**

- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/CMakeLists.txt
Comment (joiskash) [inline base/CMakeLists.txt:461]: Is this and the bottom sfml-system and sfml-audio both required?

File: base/CMakeLists.txt
Comment (joiskash) [inline base/CMakeLists.txt]: Remove this line, line 192 takes care of adding it in linux

File: base/CMakeLists.txt
Comment (joiskash) [inline base/CMakeLists.txt]: Same here

File: base/CMakeLists.txt
Comment (joiskash) [inline base/CMakeLists.txt:435]: Fix indentation

File: base/src/SoundRecord.cpp
Comment (joiskash) [inline base/src/SoundRecord.cpp:9]: Remove this if not needed

File: base/src/SoundRecord.cpp
Comment (joiskash) [inline base/src/SoundRecord.cpp:19]: Please make the indentation better.

File: base/src/SoundRecord.cpp
Comment (joiskash) [inline base/src/SoundRecord.cpp:88]: Can you please add a comment as to why this * 2 is required.

File: base/src/SoundRecord.cpp
Comment (joiskash) [inline base/src/SoundRecord.cpp:90]: Remove memcpy. Pass reference 

CC: @yashrajsapra

File: base/src/SoundRecord.cpp
Comment (joiskash) [inline base/src/SoundRecord.cpp:128]: Here you are setting everything to 0. Why?

**Review by joiskash — CHANGES_REQUESTED** [review]
Can you please add a test for "testing for dynamic change of props and validation" , if this is applicable to you

**Review by joiskash — APPROVED** [review]
Looks good to me! Good job! @kumaakh want to take a look?

@mansirthd we can just wait for Akhil, once he reviews you can merge.

File: base/include/SoundRecord.h
Comment (kumaakh) [inline base/include/SoundRecord.h:19]: Is this number of chanels (e.g. mono vs sterio) ?. If so it should be a plural

File: base/include/SoundRecord.h
Comment (kumaakh) [inline base/include/SoundRecord.h:24]: The name SoundRecord is quiet confusing.. An intuitive name like AudioCaptureSrc should be used. as proposed in #64

File: base/src/SoundRecord.cpp
Comment (kumaakh) [inline base/src/SoundRecord.cpp:5]: Interesting to see that Graphics.hpp is required, can you confirm what happens if you remove it ?

File: base/include/SoundRecord.h
Comment (kumaakh) [inline base/include/SoundRecord.h:14]: How does the user of this class know which number to be passed for "device" ? a.k.a is there a better name for this argument ?

File: base/include/SoundRecord.h
Comment (kumaakh) [inline base/include/SoundRecord.h:15]: Is this really processing rate ? I saw ithe test sets it to 200 and it looks like it is a time interval in milliseconds. In general, rate = (time interval )^-1 i.e. inverse. So let's use a better name to avoid confusing the audience. I think it is "processingInterval" in SFML, we can use "processingIntervalMS" to be clear.

File: base/test/sound_record_tests.cpp
Comment (kumaakh) [inline base/test/sound_record_tests.cpp:37]: Since we are using a wav file and we know the chanels and samplingrate, we can estimate the expected byteSize for the output file and assert that it is in the right ball park let's say +/- 2% variability allowed. it will be one step closer to automated test. so if the file came to be a grossly wrong size, no need to play it back.

File: base/test/sound_record_tests.cpp
Comment (kumaakh) [inline base/test/sound_record_tests.cpp:62]: This is a very good test, please add the assertions for output file sizes and it becomes even better

**Review by kumaakh — CHANGES_REQUESTED** [review]
added comments embedded. Most comments are about choice of words for properties and class name.

**Review by joiskash — APPROVED** [review]
Looks good, all comments have been addressed. Good job @mansirthd !

File: base/test/sound_record_tests.cpp
Comment (kumaakh) [reply base/test/sound_record_tests.cpp:48]: OK one comment which we can do it as an improvement after this merge, **the tests should always create any litter generated...** e.g.  Audiotest*.wav files generated here will be left in the developers workspace and sometimes the test will be fooled because of the litter. e.g. woru writing fails but a past copy of file exists in the same workspace. 

The solution is to add a destructor to the test  ~sound_record_tests() which deletes the files in a tolerant fashion so that if the file does not exit it does not crash and move to the next file.

**Review by kumaakh — APPROVED** [review]
raised #77  as a followup from this task

**Pull request merged**
